### ISSUE TITLE
Unified caching of expensive attributes

### ISSUE THREAD

**Issue by cgranade** [qutip/qutip#453]
Currently, ``Qobj`` has several attributes which are somewhat expensive to compute:

- ``type``
- ``isherm``
- ``istp``
- ``iscp``
- ``ishp`` (added in #452)

Of these, ``type`` and ``isherm`` are cached by ``Qobj._type`` and ``Qobj._isherm``, respectively. These should be unified, so that the other three are also cached properly. Moreover, we should add unit tests to make sure that this caching is properly invalidated under at least some reasonably common manipulations. For instance, with the caching of ``type``, it can sometimes fail in unexpected ways to build an object up then redefine its ``dims``.

**Comment by CaineArdayfio** [comment]
Hello, I would like to be assigned this if possible. I've been working on it and think I have some form of preliminary solution

**Comment by BoxiLi** [comment]
Hi @CaineArdayfio, thanks for your interest. You are very welcome to open a Pull Request if you have some solution, even not a perfect one. @Ericgig, is this still an unsolved problem?

**Comment by Ericgig** [comment]
Yes there is nothing in the code for this right now, feel free to propose your solution.

@CaineArdayfio, there is no official assignment. If, as a user, you see something that is missing, you can just propose a solution. In general, parts of Qutip that need some attention and nobody is working on:
- Documentation and notebooks are not always up to date.
- The lattice submodule is new and we have little feedback. It could be improved greatly in both easiness to use and versatility.
- Making it work better on windows.

**Comment by Ericgig** [comment]
`ishp`, `iscp` and `istp` are still not cached.

**Comment by QubitGoblin31** [comment]
Hi, I opened PR #2838  that implements caching for ishp, iscp, istp, and iscptp, with tests.

**Comment by matti-sarjala** [comment]
A small continuation path might be to reuse the useful parts of PR #2838, but scope the next PR around the cache-invalidation invariant rather than only adding cached fields.

Possible minimal acceptance criteria:

1. Cache `ishp`, `iscp`, `istp`, and `iscptp` in the same lazy style as existing cached properties.
2. Invalidate those caches when `dims`, `superrep`, or actual matrix `data` changes.
3. Avoid invalidating them for operations that only change storage format or numerical cleanup without changing the mathematical object.
4. Add regression tests for:
   - accessing the property before mutation,
   - mutating `dims` / `superrep` / `data`,
   - confirming the property is recomputed after mutation,
   - confirming copy behavior is intentional.

The key distinction seems to be:

- cache lifetime should follow the mathematical meaning of the property,
- not merely whether a setter or helper function happened to run.

That would make the next PR easier to review, especially since #2838 already identified much of the implementation surface but got stuck on review details.
